contracts/prover/LocalProver.sol

```diff
@@ -84,4 +165,257 @@ contract LocalProver is IProver, Semver {
         // Intentionally left empty as same-chain intents cannot be challenged
         // This is a no-op similar to the prove function above
     }
+
+    /**
+     * @notice Atomically fulfills an intent and escrows rewards pending secondary intent completion
+     * @dev Withdraws funds from vault, executes fulfill, stores excess in escrow.
+     *      Uses checks-effects-interactions pattern for security.
+     *      NOTE: Portal will have creator as claimant (for double-fulfill prevention),
+     *            but actual solver is tracked in escrowData.
+     * @param intentHash Hash of the intent to flash-fulfill
+     * @param intent Complete intent struct
+     * @param claimant Address of the solver eligible for rewards (stored in escrow)
+     * @param secondaryIntentHash Hash of the dependent secondary intent
+     * @param secondaryProver Prover address for the secondary intent
+     * @param secondaryDeadline Deadline of the secondary intent (for refund timing)
+     * @return results Results from the fulfill execution
+     */
+    function flashFulfill(
+        bytes32 intentHash,
+        Intent calldata intent,
+        bytes32 claimant,
+        bytes32 secondaryIntentHash,
+        address secondaryProver,
+        uint64 secondaryDeadline
+    ) external payable returns (bytes[] memory results) {
+        // CHECKS
+        require(secondaryIntentHash != bytes32(0), "Invalid secondary intent hash");
+        require(secondaryProver != address(0), "Invalid secondary prover");
+        require(_escrowedRewards[intentHash].claimant == address(0), "Already flash-fulfilled");
+
+        // EFFECTS - Create escrow first (marks as flash-fulfilled for withdrawal)
+        _escrowedRewards[intentHash] = EscrowData({
+            claimant: address(uint160(uint256(claimant))),
+            nativeAmount: 0,  // Updated after fulfill
+            tokens: new TokenAmount[](0),  // Updated after fulfill
+            secondaryIntentHash: secondaryIntentHash,
+            secondaryProver: secondaryProver,
+            secondaryDeadline: secondaryDeadline,
+            released: false
+        });
+
+        // INTERACTIONS - Withdraw to LocalProver
+        // provenIntents() now sees escrow exists, returns address(this)
+        _PORTAL.withdraw(
+            intent.destination,
+            keccak256(abi.encode(intent.route)),
+            intent.reward
+        );
+
+        // Approve tokens for fulfill execution
+        uint256 tokensLength = intent.reward.tokens.length;
+        for (uint256 i = 0; i < tokensLength; ++i) {
+            IERC20(intent.reward.tokens[i].token).safeIncreaseAllowance(
+                address(_PORTAL),
+                intent.reward.tokens[i].amount
+            );
+        }
```

Comment:
I am confused. `.fulfill` grabs the tokens defined in `.route` instead of `.reward`, so why grant allowances for the tokens in `.reward` here? Unless the assumption here is that, flash loan can only work iff reward tokens and route tokens overlap?